tools/text/msgdis.py

```diff
@@ -0,0 +1,3190 @@
+    0x6172: "NA_SE_VO_MKVO00",
+    0x6173: "NA_SE_VO_MKVO01",
+    0x6174: "NA_SE_VO_MKVO02",
+    0x6175: "NA_SE_VO_MKVO03",
+    0x6176: "NA_SE_VO_MKVO04",
+    0x6177: "NA_SE_VO_TIVO00",
+    0x6178: "NA_SE_VO_TIVO01",
+    0x6179: "NA_SE_VO_TIVO02",
+    0x617A: "NA_SE_VO_TIVO03",
+    0x617B: "NA_SE_VO_TIVO04",
+    0x617C: "NA_SE_VO_TIVO05",
+    0x617D: "NA_SE_VO_OMVO00",
+    0x617E: "NA_SE_VO_OMVO01",
+    0x617F: "NA_SE_VO_OMVO02",
+    0x6180: "NA_SE_VO_OMVO03",
+    0x6181: "NA_SE_VO_OMVO04",
+    0x6182: "NA_SE_VO_OMVO05",
+    0x6183: "NA_SE_VO_OMVO06",
+    0x6184: "NA_SE_VO_DEMO_FALL2",
+    0x6185: "NA_SE_VO_DEMO_384",
+    0x6186: "NA_SE_VO_DEMO_385",
+    0x6187: "NA_SE_VO_DEMO_386",
+    0x6188: "NA_SE_VO_DEMO_387",
+    0x6189: "NA_SE_VO_DEMO_388",
+    0x618A: "NA_SE_VO_DEMO_389",
+    0x618B: "NA_SE_VO_DEMO_390",
+    0x618C: "NA_SE_VO_DEMO_391",
+    0x618D: "NA_SE_VO_DEMO_392",
+    0x618E: "NA_SE_VO_DEMO_393",
+    0x618F: "NA_SE_VO_DEMO_394",
+}
+
+def read4(data : bytes, p : int) -> int:
+    return struct.unpack(">I", data[p:p+4])[0]
+
+def unique_or_none(lst : List[T]) -> Optional[T]:
+    if not lst:
+        return None
+    elem = lst[0]
+    for e in lst[1:]:
+        if e != elem:
+            return None
+    return elem
+
+class MessageDecoder:
+    def __init__(self, control_end : int, control_codes : Dict[int, Tuple[str, str, Optional[Tuple[Callable[[int], str]]]]], 
+                 control_header : Tuple[str, str, Optional[Tuple[Callable[[int], str]]]], extraction_charmap : Dict[int, str]) -> None:
+        self.control_end : int = control_end
+        self.control_codes : Dict[int, Tuple[str, str, Optional[Tuple[Callable[[int], str]]]]] = control_codes
+        self.control_header : Tuple[str, str, Optional[Tuple[Callable[[int], str]]]] = control_header
+        self.extraction_charmap : Dict[int, str] = extraction_charmap
+        self.msg : Optional[bytes] = None
+
+    def pop_char(self) -> int:
+        # Implement in subclass
+        raise NotImplementedError()
+
+    def pop_char_end(self) -> int:
+        # Implement in subclass
+        raise NotImplementedError()
+
+    def decode_char(self, c : int) -> str:
+        # Implement in subclass
+        raise NotImplementedError()
+
+    def pop_byte(self) -> int:
+        c, self.msg = self.msg[0], self.msg[1:]
+        return c
+
+    def pop_byte_end(self) -> int:
+        c, self.msg = self.msg[-1], self.msg[:-1]
+        return c
+
+    def pop_2byte(self) -> int:
+        u = self.pop_byte()
+        l = self.pop_byte()
+        return (u << 8) | l
+
+    def pop_2byte_end(self) -> int:
+        l = self.pop_byte_end()
+        u = self.pop_byte_end()
+        return (u << 8) | l
+
+    def format_sfx_id(self, c : int) -> str:
+        try:
+            if c & 0x800:
+                return sfx_ids[c & ~0x800]
+            else:
+                return f"{sfx_ids[c]} - SFX_FLAG"
+        except KeyError:
+            return f"0x{c:04X}"
```

Comment:
You mentioned this case in the PR description, could you list the sfx ids that this applies to?

Reply:
MessageID: SfxID
0x08E8: 0x2693
0x0963: 0x6394
0x0CEA: 0x3AE8
0x0CEF: 0x3AE8
0x0E14: 0x3AFD